src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,455 @@
+    if (!CheckJNIExceptions(env))
+        ret = (uint64_t)hash;
+
+cleanup:
+    (*env)->DeleteLocalRef(env, issuer);
+    return ret;
+}
+
+static void EnumExtensions(JNIEnv *env, jobject /*X509Certificate*/ cert, bool critical, EnumX509ExtensionsCallback cb, void *context)
+{
+    INIT_LOCALS(loc, exts, iter)
+
+    // Set<string> crit = critical ? cert.getCriticalExtensionOIDs() : cert.getNonCriticalExtensionOIDs();
+    loc[exts] = critical
+        ? (*env)->CallObjectMethod(env, cert, g_X509CertGetCriticalExtensionOIDs)
+        : (*env)->CallObjectMethod(env, cert, g_X509CertGetNonCriticalExtensionOIDs);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // Iterator<string> iter = collection.iterator();
+    // while (iter.hasNext()) {
+    //     string oid = iter.next();
+    //     byte[] data = cert.getExtensionValue(oid);
+    //     cb(oid, oid.length, data, data.length, /*critical*/ true, context);
+    // }
+    loc[iter] = (*env)->CallObjectMethod(env, loc[exts], g_SetIterator);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    jboolean hasNext = (*env)->CallBooleanMethod(env, loc[iter], g_IteratorHasNext);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    while (hasNext)
+    {
+        INIT_LOCALS(locLoop, oid, data)
+        const char *oidPtr = NULL;
+        jbyte *dataPtr = NULL;
+
+        locLoop[oid] = (*env)->CallObjectMethod(env, loc[iter], g_IteratorNext);
+        ON_EXCEPTION_PRINT_AND_GOTO(loop_cleanup);
+        jsize oidLen = (*env)->GetStringUTFLength(env, locLoop[oid]);
+        oidPtr = (*env)->GetStringUTFChars(env, locLoop[oid], NULL);
+
+        locLoop[data] = (*env)->CallObjectMethod(env, cert, g_X509CertGetExtensionValue, locLoop[oid]);
+        ON_EXCEPTION_PRINT_AND_GOTO(loop_cleanup);
+        jsize dataLen = (*env)->GetArrayLength(env, locLoop[data]);
+        dataPtr = (*env)->GetByteArrayElements(env, locLoop[data], NULL);
+
+        // X509Certificate.getExtensionValue returns the full DER-encoded data.
+        cb(oidPtr, oidLen, (const uint8_t *)dataPtr, dataLen, critical, context);
+
+        hasNext = (*env)->CallBooleanMethod(env, loc[iter], g_IteratorHasNext);
+        if (CheckJNIExceptions(env))
+            hasNext = false;
+
+    loop_cleanup:
+        if (oidPtr != NULL)
+            (*env)->ReleaseStringUTFChars(env, locLoop[oid], oidPtr);
+
+        if (dataPtr != NULL)
+            (*env)->ReleaseByteArrayElements(env, locLoop[data], dataPtr, JNI_ABORT);
+
+        RELEASE_LOCALS(locLoop, env)
+    }
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+}
+
+int32_t AndroidCryptoNative_X509EnumExtensions(jobject /*X509Certificate*/ cert, EnumX509ExtensionsCallback cb, void *context)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+
+    // [TODO] We're skipping over any errors - do we care to propagate?
+    EnumExtensions(env, cert, true /*critical*/, cb, context);
+    EnumExtensions(env, cert, false /*critical*/, cb, context);
```

Comment:
Why does this need to work in two passes?  I don't remember if we have strong ordering across Windows, Linux, and macOS, but if we do I feel like it would be in the declared order, not in declared order with criticals first... this seems to be changing that.

Reply:
The Java-based APIs only expose getting the critical and non-critical extensions separately, not just getting all extensions (at least that I could find).

I didn't see a documented guaranteed order in the .NET public APIs. I just went based on the `ReadExtensions` test, which doesn't seem to expect a specific order, since it re-orders the returned extensions:
https://github.com/dotnet/runtime/blob/0d59c78d5b40b82a3f9f617d486547444cdd2337/src/libraries/System.Security.Cryptography.X509Certificates/tests/ExtensionsTests.cs#L17-L23

Reply:
Hm... the `X509ExtensionCollection` tests seem to expect a specific order though (and assume that the `ReadExtensions` test is validating that):
https://github.com/dotnet/runtime/blob/0d59c78d5b40b82a3f9f617d486547444cdd2337/src/libraries/System.Security.Cryptography.X509Certificates/tests/CollectionTests.cs#L1336-L1338

Reply:
Android may need to do here what we do in macOS:  Ask the system to load the certificate.  Having done that, run through our own parser to get portions of the data.  Then we have the extensions in their raw format and can emit things in the .NET canonical ordering.

(I don't think we needed it for the extensions on macOS, but it was for something... so the CertificatePal object there also holds a CertData field, or something like that, which has the DER-Decoded structure)